src/main/scala/analysis/Analysis.scala

```diff
@@ -159,11 +163,10 @@ trait MemoryRegion
   * @param extent
   *   the start and end of the region
   */
-case class StackRegion(regionIdentifier: String, start: Expr, var extent: Option[RangeKey]) extends MemoryRegion:
-  override def toString: String = s"Stack(${regionIdentifier}, ${start})"
-  override def hashCode(): Int = regionIdentifier.hashCode()
+case class StackRegion(regionIdentifier: String, start: BitVecLiteral, var extent: Option[RangeKey]) extends MemoryRegion:
+  override def toString: String = s"Stack($regionIdentifier, $start)"
   override def equals(obj: Any): Boolean = obj match {
-    case StackRegion(ri, st, _) => st == start
+    case s: StackRegion => s.start == start
```

Comment:
Should we also consider the region identifier; I'm not clear what the identifier is supposed to represent. This is fine E.g. in an intra-procedural analysis if we make start the offset from SP and only compare regions in the same stack frame. In WYSINWYX they calculate all the stack regions as offsets from the SP in main() so they're comparable through their offset.

Reply:
This was just a stylistic change but the equality does need to be cleaned up, you're right. It's not good that the hashCode and equals methods don't match for these MemoryRegion classes. I'm fixing that up now.

The RegionIdentifier is just the name of the region at the moment. For a StackRegion or HeapRegion it just ends up being 'stack_1' or 'malloc_5' or something like that - just a unique identifier without any real meaning, but for a DataRegion it's the name of  whatever function or variable the region points to that is later used to resolve indirect calls, which is not really the most robust way to do that. 

Right now the analysis only tracks the stack offsets directly in relation to offsets from R31 - just tracking when there's a load(R31 + x) or store(R31 + x) (although it doesn't check the operator either), and each procedure has its own separate set of stack regions. There are obvious issues with this approach - it's only the offset relative to the current stack pointer, not relative to whatever it was at the start of the procedure or start of main, and it isn't checking the operator either.

Fixing these latter two issues will require a broader overhaul though.

Reply:
Yeah, this is a bigger change needed later I probably should have made a clearer issue about this, but its part of https://github.com/UQ-PAC/bil-to-boogie-translator/issues/72 and https://github.com/UQ-PAC/bil-to-boogie-translator/issues/73. We discussed with Yousif and Nick a while ago and decided we should track SP in abstract through procedure calls in a way that equality is defined for stack pointers originating in different procedures. Its worth noting though that Nick found that boogie verifies a lot faster if SP is given a concrete initial value in main().

Reply:
Yeah, of course we will need a way to handle & identify the same stack location being accessed from different procedures (either via a pointer as a parameter, or parameters being passed on the stack - are there any other ways it could happen?)

Make an issue for giving a concrete SP value, that's good to know. I don't know how practical that will be though, wouldn't it be necessary to set a concrete SP value for every procedure (which is obviously not really similarly practical, necessarily) to truly get the benefit? It's worth making a note of and investigating further once we have the MRA/VSA working properly, especially for programs with multiple procedures.

Reply:
> are there any other ways it could happen?

As far as VSA is concerned a stack pointer could be stored in any region, e.g. global or heap memory.

> wouldn't it be necessary to set a concrete SP value for every procedure (which is obviously not really similarly practical, necessarily) to truly get the benefit

Yeah, we can do this if a procedure doesn't have a pointer to its caller's stack and doesn't place constraints on its value, or we can keep track for the offset from main's stack and pass it through.

Reply:
Yeah I guess those cases are theoretically possible but would probably be the result of rather nasty code.

Keeping track of the offset from main's initial SP does not seem like a practical approach for programs that call the same procedure many times from different locations, or even just in cases where we have to overapproximate the number of indirect call targets.

I think the key thing will be to identify which parts of a procedure's stack can be accessed from a different procedure.